lib/Crypto/crypto_c.h

```diff
@@ -0,0 +1,302 @@
+    Sailfish_Crypto_Key_OriginDevice,
+    Sailfish_Crypto_Key_OriginSecureDevice
+};
+
+enum Sailfish_Crypto_Key_Algorithm {
+    Sailfish_Crypto_Key_AlgorithmUnknown    = 0,
+
+    Sailfish_Crypto_Key_Aes128              = 10,
+    Sailfish_Crypto_Key_Aes196,
+    Sailfish_Crypto_Key_Aes256,
+
+    Sailfish_Crypto_Key_Dsa512              = 20,
+    Sailfish_Crypto_Key_Dsa1024,
+    Sailfish_Crypto_Key_Dsa2048,
+    Sailfish_Crypto_Key_Dsa3072,
+    Sailfish_Crypto_Key_Dsa4096,
+
+    Sailfish_Crypto_Key_Rsa512              = 30,
+    Sailfish_Crypto_Key_Rsa1028,
+    Sailfish_Crypto_Key_Rsa2048,
+    Sailfish_Crypto_Key_Rsa3072,
+    Sailfish_Crypto_Key_Rsa4096,
+
+    Sailfish_Crypto_Key_NistEcc192          = 40,
+    Sailfish_Crypto_Key_NistEcc224,
+    Sailfish_Crypto_Key_NistEcc256,
+    Sailfish_Crypto_Key_NistEcc384,
+    Sailfish_Crypto_Key_NistEcc521,
+
+    Sailfish_Crypto_Key_BpEcc160            = 50,
+    Sailfish_Crypto_Key_BpEcc192,
+    Sailfish_Crypto_Key_BpEcc224,
+    Sailfish_Crypto_Key_BpEcc256,
+    Sailfish_Crypto_Key_BpEcc320,
+    Sailfish_Crypto_Key_BpEcc384,
+    Sailfish_Crypto_Key_BpEcc512
+};
+
+enum Sailfish_Crypto_Key_BlockMode {
+    Sailfish_Crypto_Key_BlockModeUnknown    = 0,
+    Sailfish_Crypto_Key_BlockModeCBC        = 1,
+    Sailfish_Crypto_Key_BlockModeCTR        = 2,
+    Sailfish_Crypto_Key_BlockModeECB        = 4,
+    Sailfish_Crypto_Key_BlockModeGCM        = 8
+};
+
+enum Sailfish_Crypto_Key_EncryptionPadding {
+    Sailfish_Crypto_Key_EncryptionPaddingUnknown    = 0,
+    Sailfish_Crypto_Key_EncryptionPaddingNone       = 1,
+    Sailfish_Crypto_Key_EncryptionPaddingPkcs7      = 2,
+    Sailfish_Crypto_Key_EncryptionPaddingRsaOaep    = 4,
+    Sailfish_Crypto_Key_EncryptionPaddingRsaOaepMgf1= 8,
+    Sailfish_Crypto_Key_EncryptionPaddingRsaPkcs1   = 16,
+    Sailfish_Crypto_Key_EncryptionPaddingAnsiX923   = 32
+};
+
+enum Sailfish_Crypto_Key_SignaturePadding {
+    Sailfish_Crypto_Key_SignaturePaddingUnknown     = 0,
+    Sailfish_Crypto_Key_SignaturePaddingNone        = 1,
+    Sailfish_Crypto_Key_SignaturePaddingRsaPss      = 2,
+    Sailfish_Crypto_Key_SignaturePaddingRsaPkcs1    = Sailfish_Crypto_Key_EncryptionPaddingRsaPkcs1,
+    Sailfish_Crypto_Key_SignaturePaddingAnsiX923    = Sailfish_Crypto_Key_EncryptionPaddingAnsiX923
+};
+
+enum Sailfish_Crypto_Key_Digest {
+    Sailfish_Crypto_Key_DigestUnknown       = 0,
+    Sailfish_Crypto_Key_DigestSha1          = 1,
+    Sailfish_Crypto_Key_DigestSha256        = 2,
+    Sailfish_Crypto_Key_DigestSha384        = 4,
+    Sailfish_Crypto_Key_DigestSha512        = 8
+};
+
+enum Sailfish_Crypto_Key_Operation {
+    Sailfish_Crypto_Key_OperationUnknown    = 0,
+    Sailfish_Crypto_Key_Sign                = 1,
+    Sailfish_Crypto_Key_Verify              = 2,
+    Sailfish_Crypto_Key_Encrypt             = 4,
+    Sailfish_Crypto_Key_Decrypt             = 8
+};
+
+struct Sailfish_Crypto_Key {
+    struct Sailfish_Crypto_Key_Identifier *identifier;
```

Comment:
Padding only makes sense if this structure is supposed to be allocated by the client of the library. If it's only allocated by the library itself then padding is unnecessary - old clients just won't know that the structure is bigger than they thought and won't access the new fields. Direct access to the public fields is easier, more readable and more efficient than a function call, I'm personally fine with that (and do that myself quite often). Obviously, new fields have to be added to the end of the structure.

Reply:
You're right for the padding. GObject is using it because it is inserting parent struct in child struct and they need to keep the offset right for child when adding new attributes in parent. It's not the case here. So no padding!

Reply:
My only real opinion about this is: we shouldn't commit to supporting this C API with BC/SC guarantees until we have real clients using the C++ API (including things which currently aren't fleshed out, but are very important, like certificates), because currently I see this API as being entirely WIP / subject to change.

Basically, this API exists as a prototype only and I expect it to evolve over the coming weeks as we actually start to implement real use-cases with the APIs.

Aside from that, my experience with pure C APIs is obviously quite limited compared to my experience with C++ APIs so I will no doubt have made a lot of mistakes / many improvements can be made / patches welcome ;-)